Hunting and deforestation support.

zhaodice opened the pull request:
## Description

implement woods or meat dropped when attacking tree or animal

## Issues fixed by this PR

<!--- Put the links of issues that may be fixed by this PR here (if any). -->
#1071 

## Type of changes

<!--- Put an `x` in all the boxes that apply your changes. -->

- [ ] Bug fix
- [x] New feature 
- [x] Enhancement
- [ ] Documentation

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] My pull request is unique and no other pull requests have been opened for these changes
- [x] I have read the [Contributing note](https://github.com/Grasscutters/Grasscutter/blob/stable/CONTRIBUTING.md) and [Code of conduct](https://github.com/Grasscutters/Grasscutter/blob/development/CODE_OF_CONDUCT.md)
- [x] I am responsible for any copyright issues with my code if it occurs in the future.

zhaodice committed:
add drops for animals wild pig,fishes,foxes,birds

zhaodice committed:
append fox

zhaodice committed:
Deforestation Support

implement drop woods when attacking tree

zhaodice committed:
Deforestation support (remove prints)

implement drop woods when attacking tree

zhaodice committed:
Deforestation support (remove prints)

implement drop woods when attacking tree

Melledy commented on src/main/java/emu/grasscutter/server/packet/recv/HandlerHitTreeNotify.java:
```diff
@@ -0,0 +1,65 @@
+
+@Opcodes(PacketOpcodes.HitTreeNotify)
+public class HandlerHitTreeNotify extends PacketHandler {
+    private final HashMap<Integer,HashMap<Integer,HitTreeRecord>> hitRecord = new HashMap<>();
```
`hitRecord` should be stored as a transient (so it doesnt get saved to db) hashmap in the Player object. That way, you avoid a memory leak when players stop hitting trees and log out.

zhaodice replied:
> `hitRecord` should be stored as a transient (so it doesnt get saved to db) hashmap in the Player object. That way, you avoid a memory leak when players stop hitting trees and log out.

I see, so I decide to use `AutoRecycleHashMap`, It will auto recycle key and value

zhaodice replied:
> `hitRecord` should be stored as a transient (so it doesnt get saved to db) hashmap in the Player object. That way, you avoid a memory leak when players stop hitting trees and log out.

There may have some problems, please wait my checking

zhaodice replied:
> `hitRecord` should be stored as a transient (so it doesnt get saved to db) hashmap in the Player object. That way, you avoid a memory leak when players stop hitting trees and log out.

All done, please review again.

zhaodice committed:
add AutoRecycleHashMap

Map's KEY is automatic expire if key long time no use (query or modify from HashMap)

zhaodice committed:
use AutoRecycleHashMap in case of memory leak

zhaodice committed:
fix bug

zhaodice committed:
remove prints

zhaodice commented:
I want to confirm that `public class HandlerHitTreeNotify extends PacketHandler {` is Singleton Mode(only new one instance). if true, that is ok.

zhaodice committed:
static AutoRecycleHashMap

zhaodice commented:
static AutoRecycleHashMap could ensure single HashMap instance in case of memory leak when mult-new HandlerHitTreeNotify without call HandlerHitTreeNotify.clear().

GanyusLeftHorn commented on src/main/java/emu/grasscutter/server/packet/recv/HandlerHitTreeNotify.java:
```diff
@@ -0,0 +1,69 @@
+@Opcodes(PacketOpcodes.HitTreeNotify)
+public class HandlerHitTreeNotify extends PacketHandler {
+    private final static int RECORD_EXPIRED_SECONDS = 60*5;
+    private final static AutoRecycleHashMap<Integer,HashMap<Integer,HitTreeRecord>> hitRecord = new AutoRecycleHashMap<>(RECORD_EXPIRED_SECONDS);
```
Why don't you just move this to `Player`, as Melledy had suggested? This `AutoRecycleHashMap` just adds a lot of unnecessary complexity.

zhaodice replied:
> Why don't you just move this to `Player`, as Melledy had suggested? This `AutoRecycleHashMap` just adds a lot of unnecessary complexity.

all done

GanyusLeftHorn commented on src/main/java/emu/grasscutter/server/packet/recv/HandlerHitTreeNotify.java:
```diff
@@ -0,0 +1,69 @@
+    public void handle(GameSession session, byte[] header, byte[] payload) throws Exception {
+        HitCollision hit = HitCollision.parseFrom(payload);
+        VectorOuterClass.Vector hitPosition = hit.getHitPoint();
+        int woodType = hit.getHitColliderTypeValue();
+        if(woodType>=1 && woodType<=12) {// is a available wood type
```
I don't see how this is supposed to work. `HitColliderType` does not tell you the type of wood you hit. It, as the name suggests, tells you the type of collider/hitbox.

zhaodice replied:
> I don't see how this is supposed to work. `HitColliderType` does not tell you the type of wood you hit. It, as the name suggests, tells you the type of collider/hitbox.

but I found there is regular pattern that ColliderTypeValue is 1 to 12, each is in 101301,101302,101303..etc,Do you have a better way to know which type I hit wood?

GanyusLeftHorn replied:
I don't see how that would be possible. The return value of `getHitColliderTypeValue()` should only ever take values from -1 to 3, corresponding to the enum values defined in `emu.grasscutter.net.proto.HitColliderTypeOuterClass.HitColliderType`. And when I tested your code while doing my review, that is exactly the behavior I got. In fact, the value returned was always zero.

zhaodice replied:
> I don't see how that would be possible. The return value of `getHitColliderTypeValue()` should only ever take values from -1 to 3, corresponding to the enum values defined in `emu.grasscutter.net.proto.HitColliderTypeOuterClass.HitColliderType`. And when I tested your code while doing my review, that is exactly the behavior I got. In fact, the value returned was always zero.

Oh, I see, I used the wrong proto file, but it happened to resolve the type of wood. I'll re analyze the protocol!

GanyusLeftHorn commented on src/main/java/emu/grasscutter/server/packet/recv/HandlerHitTreeNotify.java:
```diff
@@ -0,0 +1,69 @@
+                if (currentRecord.containsKey(positionHash)) {
+                    long currentTime = System.currentTimeMillis();
+                    HitTreeRecord record = currentRecord.get(positionHash);
+                    if (currentTime - record.time > RECORD_EXPIRED_SECONDS *1000) {// fresh wood after 5 min
+                        record.times = 1;
+                        record.time = currentTime;
+                    } else {
+                        if (record.times >= 3) {// too many wood they have deforested, no more wood dropped!
+                            return;
+                        } else {
+                            record.times++;
+                        }
+                    }
```
On the live server, wood also respawns 
(a) if you hit ten other trees since the last time you hit this particular tree, and 
(b) if you log out and log in again. 
Maybe we should try and replicate that? (b) should automatically be handled if you move the hit information to `Player`.

zhaodice committed:
fix problems

zhaodice committed:
Delete AutoRecycleHashMap.java

zhaodice committed:
remove log

zhaodice committed:
fix build

Birdulon commented on src/main/java/emu/grasscutter/game/managers/DeforestationManager/HitTreeRecord.java:
```diff
@@ -0,0 +1,77 @@
+    private final static int RECORD_MAX_TIMES_OTHER_HIT_TREE = 10; // if hit 10 times other trees, reset wood
+
+    private final int unique;
+    private final ArrayList<Integer> otherTrees = new ArrayList<>(); // hit other tree
```
This would probably be better off as a global list that stores the last 10 trees that have been hit, as a circular buffer. It looks like this stores every tree that has been hit quadratically in this current form, which seems like a bad idea.

Birdulon commented on src/main/java/emu/grasscutter/game/managers/DeforestationManager/DeforestationManager.java:
```diff
@@ -0,0 +1,90 @@
+                HitTreeRecord record = searchRecord(positionHash);
+                if (record == null) {
+                    record = new HitTreeRecord(positionHash);
+                    currentRecord.add(record);
```
A simple fix might be adding `if (currentRecord.size() > 10) {currentRecord.remove(0);}` after this and deleting all the `recordOtherTree` logic

zhaodice replied:
> A simple fix might be adding `if (currentRecord.size() > 10) {currentRecord.remove(0);}` after this and deleting all the `recordOtherTree` logic

all done

zhaodice committed:
improve

Birdulon commented:
Deforestation works on my machine, haven't seen any meat drop from boars or small birds yet.

zhaodice commented:
> Deforestation works on my machine, haven't seen any meat drop from boars or small birds yet.

delete /data/drop.json in order to update new one.

memetrollsXD commented:
hahaha deforestation support

zhaodice commented:
> hahaha deforestation support

It sounds I am a bad girl :(

Birdulon commented:
> > Deforestation works on my machine, haven't seen any meat drop from boars or small birds yet.
> 
> delete /data/drop.json in order to update new one.

aha, completely forgot about that

arub3do commented on src/main/java/emu/grasscutter/server/packet/recv/HandlerHitTreeNotify.java:
```diff
@@ -0,0 +1,22 @@
+
+/**
+ * Implement Deforestation Function
+ * @author 赵怡然(Github zhaodice)
```
```suggestion
```
Be careful, or you need a force push.

zhaodice replied:
> Be careful, or you need a force push.

this is a virtual name,don't worry

zhaodice committed:
remove unnecessary information

Co-authored-by: Albedo <105265570+arub3do@users.noreply.github.com>
